docs/DesignDocs/DebuggerSupportForMultiModuleEnclaves.md

```diff
@@ -0,0 +1,268 @@
+
+- [feature.openlibos](https://github.com/openenclave/openenclave/tree/feature.openlibos)
+
+  This branch of OE SDK runs a library OS within an enclave. It also supports
+  enclave comprised of multiple executables and shared libraries.
+  https://github.com/openenclave/openenclave/blob/feature.openlibos/debugger/pythonExtension/gdb_sgx_plugin.py
+
+- [feature/dynamic_binding](https://github.com/openenclave/openenclave/tree/feature/dynamic_binding)
+
+  This branch of OE SDK allows loading and running multi-module enclaves with
+  the goal of permitting use of FIPS certified crypto modules. It also supports
+  debugging multi-module enclaves:
+  https://github.com/openenclave/openenclave/commit/254ae94324b619861f58d0110a1affac37e5a2a4?branch=254ae94324b619861f58d0110a1affac37e5a2a4&diff=split
+
+
+The debugger changes required to support multi-module enclaves are quite similar
+in the above three implementations. The following proposal formally incorporates
+the necessary debugger changes into the Debugger Contract.
+
+Debugger Contract Extensions to Support Multi-Module Enclaves
+-------------------------------------------------------------
+
+The following `oe_debug_module_t` structure captures the information necessary
+to debug a module:
+
+```c
+
+    typedef struct _debug_module_t
+	{
+	    // Magic value at start of the structure.
+		uint64_t magic;
+		// Structure version.
+		uint64_t version;
+
+		// Path of module's ELF binary.
+		char* path;
+		uint64_t path_length;
+
+		// Memory region where the enclave has been loaded.
+		uint64_t base_address;
+		uint64_t size;
+	} oe_debug_module_t;
+	
+```
+
+The `path` and `path_length` together specify the path of the module's ELF
+binary. The `base_address` and `size` together specify the memory region
+where the enclave has been loaded.
+
+The following notification functions are used to notify the debugger about the
+loading and unloading of a module.
+
+- `oe_result_t oe_result_t oe_debug_notify_module_loaded(oe_debug_module_t* module)`
+
+   This function is called by the SDK's host-runtime after a module has been
+   loaded, but before it has been initialized. A module is initialized by invoking
+   it's constructor functions. It doesn not matter whether the module has been relocated
+   or not before the debugger is notified about the module.
+
+   In situations where a module may be dynamically loaded by an enclave, as is the case
+   when SGX LKL/openlibos launch applications within the enclave, this notification
+   function is called by the SDK's enclave-side runtime prior to the initialization
+   of the module.
+
+- `oe_result_t oe_debug_notify_module_unloaded(oe_debug_module_t* module)`
+
+   This function is called by the SDK runtimes (host-side or enclave-side), after a
+   module has been unloaded.
+
+Debugger Handling of Module Load/Unload Notifications
+
+- Linux
+  Similar to enclave load/unload, the debugger sets internal breakpoints in
```

Comment:
From the SIG-arch meeting, there was a discussion as to whether this applied to all loaded modules, including the primary enclave app module, and the expectation was that it would exclude the primary enclave module (which would have the equivalent tasks handled as part of the enclave start notification).

The feedback was that it would be nice if the primary enclave could be handled uniformly, with whatever additional tasks specific to enclave start (e.g. TCS handling) being factored out into a separate implementation that is composable with the module load. There was some concern that this could break the existing contract with the windbg, and that would need to be investigated.

Reply:
While, at first glance, it may seem nicer to treat primary and secondary modules the same, and factor out extra details of the primary module into a separate notification, it does not seem that much valuable to me for the following reasons:
- We are losing data in this abstraction. The fact that a module is a primary module now has to be figured out in another way. 
- An enclave that does not use multiple modules will now require extra notifications to enable debugging where previously it used to do that with fewer notifications.
- The information about the primary module will now be sent via multiple notifications instead of being atomically sent in a single notification. I suspect this could make it difficult for the debugger to correlate information if multiple threads were simultaneously creating enclaves.
- This change is a breaking change that will require a change to all the debuggers (oegdb, VS, WinDbg) to debug single module enclaves. Additionally, it might also require work in Intel SDK to preserve backwards compatibility. It is not clear to me what the bar is for justifying breaking changes in published contracts.